pkg/plugin/velero/item_snapshotter/v1/item_snapshotter.go

```diff
@@ -0,0 +1,133 @@
+type ProgressInput struct {
+	// ItemID is the id of item that was stored in the backup
+	ItemID velero.ResourceIdentifier
+	// SnapshotID is the snapshot ID returned by ItemSnapshotter
+	SnapshotID string
+	// Backup is the representation of the restore resource processed by Velero.
+	Backup *api.Backup
+}
+
+type ProgressOutput struct {
+	Completed                       bool      // True when the operation has completed, either successfully or with a failure
+	Err                             error     // Set when the operation has failed
+	ItemsCompleted, ItemsToComplete int64     // The number of items that have been completed and the items to complete
+	Started, Updated                time.Time // When the upload was started and when the last update was seen.  Time 0 (time.Unix(0, 0)) is returned if unknown.
+}
+
+type DeleteSnapshotInput struct {
+	SnapshotID       string
+	ItemFromBackup   runtime.Unstructured
+	SnapshotMetadata map[string]string
+	Params           map[string]string
+}
+
+type CreateItemInput struct {
+	SnapshottedItem  runtime.Unstructured
+	SnapshotID       string
+	ItemFromBackup   runtime.Unstructured
+	SnapshotMetadata map[string]string
+	Params           map[string]string
+	// Restore is the representation of the restore resource processed by Velero.
+	Restore *api.Restore
+}
+
+type CreateItemOutput struct {
+	// UpdatedItem is the item being restored mutated by ItemAction.
+	UpdatedItem runtime.Unstructured
+
+	// AdditionalItems is a list of additional related items that should
+	// be restored.
+	AdditionalItems []velero.ResourceIdentifier
+
+	// SkipRestore tells velero to stop executing further actions
+	// on this item, and skip the restore step. When this field's
+	// value is true, AdditionalItems will be ignored.
+	SkipRestore bool
+}
+
+// ItemSnapshotter handles snapshots on an individual item being backed up.
+type ItemSnapshotter interface {
+
+	// Init prepares the ItemSnapshotter for usage using the provided map of
+	// configuration key-value pairs. It returns an error if the ItemSnapshotter
+	// cannot be initialized from the provided config.
+	Init(config map[string]string) error
+
+	// AppliesTo returns information about which resources this action should be invoked for.
+	// An ItemSnapshotter's SnapshotItem method will only be invoked on items that match the returned
+	// selector. A zero-valued ResourceSelector matches all resources.
+	AppliesTo() (velero.ResourceSelector, error)
+
+	// AlsoHandles is called for each item this ItemSnapshotter should handle and returns any items
+	// which will be handled by this plugin when snapshotting the item.  For example, a database may expose
+	// a resource that can be snapshotted. If the database uses a PVC that will be snapshotted/backed up as
+	// part of the database snapshot, that PVC should be returned when AlsoHandles is invoked
+	AlsoHandles(input *AlsoHandlesInput) ([]velero.ResourceIdentifier, error)
```

Comment:
This returns the same result as `Additionalitems` in `SnapshotItemOutput`?

Reply:
I updated the comment to explain:

	// AlsoHandles is called for each item this ItemSnapshotter should handle and returns any items
	// which will be handled by this plugin when snapshotting the item.  These items will be excluded from the
	// items being backed up.  AlsoHandles will be called before SnapshotItem is called.  For example, a database may expose
	// a database resource that can be snapshotted. If the database uses a PVC that will be snapshotted/backed up as
	// part of the database snapshot, that PVC should be returned when AlsoHandles is invoked.  This is different from
	// AdditionalItems (returned in SnapshotItemOutput and CreateItemOutput) which are specifying additional resources
	// that Velero should store in the backup or create.

Reply:
I think we should be explicit in the comment here to explain if `AlsoHandles` returns the direct children PEs only, or does it recursively traverse the subtree to return all leaf items.

Reply:
It will return everything that should be excluded from the caller's list of things to do.  If Astrolabe is backing the ItemSnapshotter, the expectation is that it returns all of the PEs that will be snapshotted so they can be removed from the list.  ItemSnapshotter is not limited to Astrolabe, however, and is not a Protected Entity interface here.

Reply:
So all children are handled in one call of `SnapshotItem`, right? The resources returned by `AlsoHandles ` are only used to tell the caller to just ignore them

Reply:
Yes, that's the plan